### PR TITLE
Add an instant post search	

### PR DESCRIPTION
**Pull request by javiereguiluz**
I've taken the code commitedd by @voronkovich in #173 and made some changes to it. Let's see if you like it. I propose to implement it like Google search results instead of displaying it in the main menu:

search-results

The reason is that the main menu is crowded when the user is logged in and the search bar looks bad.

### CODE REVIEW

File: src/AppBundle/Repository/PostRepository.php [hunk 1]
```diff
@@ -59,4 +59,51 @@ private function createPaginator(Query $query, $page)
 
         return $paginator;
     }
+
+    public function findBySearchQuery($rawQuery, $limit = Post::NUM_ITEMS)
```
Comment (voronkovich) [inline src/AppBundle/Repository/PostRepository.php]: Missing docblock

File: src/AppBundle/Repository/PostRepository.php [hunk 2]
```diff
@@ -59,4 +59,51 @@ private function createPaginator(Query $query, $page)
...
+        }
+
+        return $queryBuilder
+            ->orderBy('p.publishedAt', 'DESC')
+            ->setMaxResults($limit)
+            ->getQuery()
+            ->getResult();
+    }
+
+    private function sanitizeSearchQuery($query)
```
Comment (voronkovich) [inline src/AppBundle/Repository/PostRepository.php]: Missing docblock

File: src/AppBundle/Controller/BlogController.php [hunk 3]
```diff
@@ -150,4 +151,31 @@ public function commentFormAction(Post $post)
             'form' => $form->createView(),
         ]);
     }
+
+    /**
+     * @Route("/search", name="blog_search")
+     * @Method("GET")
+     *
+     * @return JsonResponse
```
Comment (yceruto) [inline src/AppBundle/Controller/BlogController.php]: `JsonResponse` inherits from `Response` so `JsonResponse` should be enough.

File: src/AppBundle/Controller/BlogController.php [hunk 4]
```diff
@@ -150,4 +151,31 @@ public function commentFormAction(Post $post)
...
+        $results = [];
+        foreach ($posts as $post) {
+            $results[] = [
+                'title' => htmlspecialchars($post->getTitle()),
+                'summary' => htmlspecialchars($post->getSummary()),
+                'url' => $this->generateUrl('blog_post', ['slug' => $post->getSlug()]),
+            ];
+        }
+
+        return new JsonResponse($results);
```
Comment (yceruto) [inline src/AppBundle/Controller/BlogController.php]: `return $this->json($results)`?

File: src/AppBundle/Controller/BlogController.php (hunk 3 above)
Comment (voronkovich) [inline src/AppBundle/Controller/BlogController.php]: @yceruto, the controller returns a `Response` instance when a request is not an `XMLHttpRequest`, so I think `JsonResponse` is not enough for this case.

File: app/Resources/views/blog/search.html.twig [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+{% block main %}
+    <form action="{{ path('blog_search') }}" method="get">
+        <div class="form-group">
+            <input name="q" type="text" class="form-control search-field" placeholder="{{ 'post.search_for'|trans }}" autocomplete="off" autofocus>
+        </div>
+    </form>
+
+    <div id="results">
+    </div>
+{% endblock %}
```
Comment (voronkovich) [inline app/Resources/views/blog/search.html.twig]: @javiereguiluz, you forgot to add a widget with the `Show source` button.

File: src/AppBundle/Repository/PostRepository.php [hunk 6]
```diff
@@ -59,4 +59,63 @@ private function createPaginator(Query $query, $page)
...
+
+        return $queryBuilder
+            ->orderBy('p.publishedAt', 'DESC')
+            ->setMaxResults($limit)
+            ->getQuery()
+            ->getResult();
+    }
+
+    /**
+     * Removes all non-alphanumeric characters except whitespaces
```
Comment (voronkovich) [inline src/AppBundle/Repository/PostRepository.php]: Missing dot at the end.

File: src/AppBundle/Repository/PostRepository.php [hunk 7]
```diff
@@ -59,4 +59,63 @@ private function createPaginator(Query $query, $page)
 
         return $paginator;
     }
+
+    /**
+     * @param string $rawQuery The search query as input by the user
+     * @param int $limit       The maximum number of results returned
```
Comment (voronkovich) [inline src/AppBundle/Repository/PostRepository.php]: Incorrect indent.

**Comment by javiereguiluz** [comment]
@voronkovich thanks for contributing this feature and thanks to reviewers too!

**Pull request merged**